File: FE/controlflow.cc
#include "FE/controlflow.h"

#include "FE/eval.h"

namespace cwerg::fe {

bool CondIsAlwaysTrue(Node cond) { return Node_x_eval(cond) == kConstTrue; }

bool CondIsAlwaysFalse(Node cond) { return Node_x_eval(cond) == kConstFalse; }

bool ContainsBreakRecurtsively(Node node_list, Node block_target) {
  for (Node n = node_list; !n.isnull(); n = Node_next(n)) {
    switch (n.kind()) {
      case NT::StmtBreak:
        return Node_x_target(n) == block_target;
      case NT::StmtReturn:
      case NT::StmtTrap:
      case NT::StmtContinue:
        return false;
      case NT::StmtIf:
        if (!CondIsAlwaysTrue(Node_cond(n))) {
          if (ContainsBreakRecurtsively(Node_body_f(n), block_target))
            return true;
        }
        if (!CondIsAlwaysFalse(Node_cond(n))) {
          if (ContainsBreakRecurtsively(Node_body_t(n), block_target))
            return true;
        }
        break;
      case NT::StmtDefer:
        return ContainsBreakRecurtsively(Node_body(n), block_target);
      case NT::StmtCond: {
        for (Node c = Node_cases(n); !c.isnull(); c = Node_next(c)) {
          if (ContainsBreakRecurtsively(Node_body(c), block_target)) {
            return true;
          }
        }
      }
      default:
        break;
    }
  }
  return false;
}

bool LastStmtIsBreak(Node body, Node block_target) {
  if (body.isnull()) return false;
  Node last = NodeLastSibling(body);
  if (last.kind() == NT::StmtBreak) {
    return Node_x_target(last) == block_target;
  }
  return false;
}

// forward declaration needed because of mutual recursion
bool BodyFallsThru(Node body);

bool LastStmtFallsThrough(Node last) {
  switch (last.kind()) {
    case NT::StmtReturn:
    case NT::StmtTrap:
    case NT::StmtContinue:
      return false;
    case NT::StmtBreak:
      return true;
    case NT::StmtIf: {
      if (!CondIsAlwaysTrue(Node_cond(last))) {
        if (BodyFallsThru(Node_body_f(last))) return true;
      }
      if (!CondIsAlwaysFalse(Node_cond(last))) {
        if (BodyFallsThru(Node_body_t(last))) return true;
      }
      return false;
    }
    case NT::StmtCond: {
      bool seen_always_true = false;

      for (Node c = Node_cases(last); !c.isnull(); c = Node_next(c)) {
        if (CondIsAlwaysTrue(Node_cond(c))) seen_always_true = true;
        if (BodyFallsThru(Node_body(c))) {
          return true;
        }
      }
      return !seen_always_true;
    }
    case NT::StmtBlock: {
      Node body = Node_body(last);
      if (body.isnull()) return true;
      if (ContainsBreakRecurtsively(body, last)) return true;
      return LastStmtFallsThrough(NodeLastSibling(body));
    }
    default:
      return true;
  }
}

bool BodyFallsThru(Node body) {
  if (body.isnull()) return true;
  return LastStmtFallsThrough(NodeLastSibling(body));
}

void ModVerifyFunFallthrus(Node mod) {
  ASSERT(mod.kind() == NT::DefMod, "");
  for (Node fun = Node_body_mod(mod); !fun.isnull(); fun = Node_next(fun)) {
    if (fun.kind() != NT::DefFun || Node_has_flag(fun, BF::EXTERN) ||
        CanonType_is_void(CanonType_result_type(Node_x_type(fun)))) {
      continue;
    }
    if (BodyFallsThru(Node_body(fun))) {
      CompilerError(Node_srcloc(fun))
          << "function may fall thru without returning";
    }
  }
}

struct Scope {
  Scope(Node anode) : target(anode) {}
  Node target;
  std::vector<Node> defer_stmts;
};

Node EliminateDeferRecursively(Node node, std::vector<Scope>* scopes) {
  auto handle_cfg = [scopes](Node target, NodeChain* chain) {
    for (auto s = scopes->rbegin(); s != scopes->rend(); ++s) {
      for (auto d = s->defer_stmts.rbegin(); d != s->defer_stmts.rend(); ++d) {
        for (Node n = Node_body(*d); !n.isnull(); n = Node_next(n)) {
          NodeToNodeMap dummy1;
          NodeToNodeMap dummy2;
          chain->Append(NodeCloneRecursively(n, &dummy1, &dummy2));
        }
      }
      if (s->target == target) break;
    }
  };

  if (node.kind() == NT::StmtDefer) {
    scopes->back().defer_stmts.push_back(node);
  }

  if (NodeHasField(node, NFD_X_FIELD::target)) {
    NodeChain chain;
    handle_cfg(Node_x_target(node), &chain);
    chain.Append(node);
    return chain.First();
  }

  auto& core = gNodeCore[node];
  for (int i = 0; i < MAX_NODE_CHILDREN; ++i) {
    Node child = core.children_node[i];
    if (!NodeIsNode(child)) continue;
    bool is_new_scope =
        (i == SLOT_BODY) || (Node_kind(node) == NT::StmtIf && i == SLOT_BODY_T);
    if (is_new_scope) {
      scopes->emplace_back(Scope(node));
    }

    NodeChain new_children;
    do {
      Node next = Node_next(child);
      Node_next(child) = kNodeInvalid;
      Node new_child = EliminateDeferRecursively(child, scopes);
      new_children.Append(new_child);
      child = next;
    } while (!child.isnull());

    if (is_new_scope) {
      Node last_child = new_children.Last();
      const Scope& current = scopes->back();
      if (last_child.isnull() ||
          !NodeHasField(last_child, NFD_X_FIELD::target)) {
        handle_cfg(current.target, &new_children);
      }
      for (auto d = current.defer_stmts.begin(); d != current.defer_stmts.end();
           ++d) {
        NodeFreeRecursively(*d);
      }
      scopes->pop_back();
    }
    core.children_node[i] = new_children.First();
  }

  if (node.kind() == NT::StmtDefer) {
    return kNodeInvalid;
  }
  return node;
}

void FunEliminateDefer(Node fun) {
  std::vector<Scope> scope;
  EliminateDeferRecursively(fun, &scope);
}

void FunAddMissingReturnStmts(Node fun) {
  CanonType result_ct = CanonType_result_type(Node_x_type(fun));
  CanonType unwrapped_ct = CanonType_get_unwrapped(result_ct);
  if (!CanonType_is_void(unwrapped_ct)) {
    return;
  }
  Node last = NodeLastSibling(Node_body(fun));
  if (last.kind() == NT::StmtReturn) {
    return;
  }
  const SrcLoc& sl = last.isnull() ? Node_srcloc(fun) : Node_srcloc(last);
  Node void_val = NodeNew(NT::ValVoid);
  NodeInitValVoid(void_val, kStrInvalid, sl, result_ct);

  Node ret = NodeNew(NT::StmtReturn);
  NodeInitStmtReturn(ret, void_val, kStrInvalid, sl, fun);
  if (last.isnull()) {
    Node_body(fun) = ret;
  } else {
    Node_next(last) = ret;
  }
}

void FunCanonicalizeRemoveStmtCond(Node fun) {
  auto replacer = [](Node node, Node parent) -> Node {
    if (node.kind() != NT::StmtCond) return node;
    std::vector<Node> cases;
    for (Node c = Node_cases(node); !c.isnull(); c = Node_next(c)) {
      cases.push_back(c);
    }
    if (cases.empty()) return kNodeInvalid;

    Node out = kNodeInvalid;
    for (auto it = cases.rbegin(); it != cases.rend(); ++it) {
      if (CondIsAlwaysTrue(Node_cond(*it))) {
        ASSERT(out == kNodeInvalid,
               "'case true' shold be the last case in a cond statement");
        out = Node_body(*it);
        NodeFree(Node_cond(*it));
        NodeFree(*it);
      } else {
        Node t = NodeNew(NT::StmtIf);
        NodeInitStmtIf(t, Node_cond(*it), Node_body(*it), out, kStrInvalid,
                       Node_srcloc(*it));
        NodeFree(*it);
        out = t;
      }
    }
    NodeFree(node);
    return out;
  };
  MaybeReplaceAstRecursivelyPost(fun, replacer, kNodeInvalid);
}

}  // namespace cwerg::fe

File: FE/controlflow.py
from typing import Any, Optional

from FE import cwast
from FE import type_corpus
from FE import eval
from FE import canonicalize


def CondIsAlwaysTrue(cond: Any) -> bool:
    return cond.x_eval == eval.VAL_TRUE


def CondIsAlwaysFalse(cond: Any) -> bool:
    return cond.x_eval == eval.VAL_FALSE


def _ContainsBreakRecurtsively(node_list: list[Any], block_target: cwast.StmtBlock) -> bool:
    for node in node_list:
        if isinstance(node, cwast.StmtBreak):
            # no reason to continue - reamaining stmts are unreachable
            return node.x_target is block_target
        elif isinstance(node, (cwast.StmtReturn, cwast.StmtTrap, cwast.StmtContinue)):
            return False
        elif isinstance(node, cwast.StmtBlock):
            return _ContainsBreakRecurtsively(node.body, block_target)
        elif isinstance(node, cwast.StmtIf):
            if not CondIsAlwaysTrue(node.cond):
                if _ContainsBreakRecurtsively(node.body_f, block_target):
                    return True
            if not CondIsAlwaysFalse(node.cond):
                if _ContainsBreakRecurtsively(node.body_t, block_target):
                    return True
        elif isinstance(node, cwast.StmtDefer):
            return _ContainsBreakRecurtsively(node.body, block_target)
        elif isinstance(node, cwast.StmtCond):
            for c in node.cases:
                if _ContainsBreakRecurtsively(c.body, block_target):
                    return True
    return False


def _LastStmtFallsThrough(last) -> bool:
    if isinstance(last, (cwast.StmtContinue, cwast.StmtReturn, cwast.StmtTrap)):
        return False
    elif isinstance(last, cwast.StmtBreak):
        return True
    elif isinstance(last, cwast.StmtIf):
        if not CondIsAlwaysTrue(last.cond):
            if _BodyFallsThru(last.body_f):
                return True
        if not CondIsAlwaysFalse(last.cond):
            if _BodyFallsThru(last.body_t):
                return True
        return False
    elif isinstance(last, cwast.StmtCond):
        seen_always_true = False
        for x in last.cases:
            if CondIsAlwaysTrue(x.cond):
                seen_always_true = True
            # This is incomplete as it does not ensure full coverge of all connditions
            if _BodyFallsThru(x.body):
                return True
        return not seen_always_true
    elif isinstance(last, cwast.StmtBlock):
        if len(last.body) == 0:
            return True
        elif len(last.body) >= 2:
            if _ContainsBreakRecurtsively(last.body, last):
                return True
        return _LastStmtFallsThrough(last.body[-1])
    else:
        return True


def _BodyFallsThru(body) -> bool:
    if len(body) == 0:
        return True
    return _LastStmtFallsThrough(body[-1])


def ModVerifyFunFallthrus(mod: cwast.DefMod):
    for fun in mod.body_mod:
        if not isinstance(fun, cwast.DefFun) or fun.extern or fun.x_type.result_type().is_void():
            continue
        if _BodyFallsThru(fun.body):
            # from FE import pp_ast
            # import sys
            # pp_ast.DumpNode(fun, fout=sys.stdout)
            cwast.CompilerError(
                fun.x_srcloc, f"{fun.name} has non-void return type but does not end with a return statement")


# TODO: try converting this to VisitAstRecursivelyPreAndPost
def _EliminateDeferRecursively(node: Any, scopes: list[tuple[Any, list[Any]]]):

    def handle_cfg(target):
        out = []
        for scope, defers in reversed(scopes):
            for defer in reversed(defers):
                clone = cwast.CloneNodeRecursively(defer, {}, {})
                out += clone.body
            if scope is target:
                break
        for x in out:
            assert not isinstance(x, cwast.StmtDefer)
        return out

    if isinstance(node, cwast.StmtDefer):
        scopes[-1][1].append(node)

    if cwast.NF.TARGET_ANNOTATED in node.FLAGS:
        # inject the defer bodies just before the control flow change
        return handle_cfg(node.x_target) + [node]

    for nfd in node.__class__.NODE_FIELDS:
        field = nfd.name
        if nfd.kind is cwast.NFK.NODE:
            child = getattr(node, field)
            new_child = _EliminateDeferRecursively(child, scopes)
            assert new_child is None
            if new_child:
                assert not isinstance(new_child, list)
                setattr(node, child, new_child)
        else:
            if field in cwast.NEW_SCOPE_FIELDS:
                scopes.append((node, []))
            #
            children = getattr(node, field)
            new_children = []
            change = False
            for n, child in enumerate(children):
                new_child = _EliminateDeferRecursively(child, scopes)
                if new_child is None:
                    new_children.append(child)
                else:
                    assert isinstance(new_child, list)
                    new_children += new_child
                    change = True

            #
            if field in cwast.NEW_SCOPE_FIELDS:
                # handle end of scope if the last statement was NOT a control flow change
                if not new_children or cwast.NF.TARGET_ANNOTATED not in children[-1].FLAGS:
                    out = handle_cfg(scopes[-1][0])
                    if out:
                        new_children += out
                        change = True
                scopes.pop(-1)
            #
            if change:
                setattr(node, field, new_children)

    if isinstance(node, cwast.StmtDefer):
        return []
    return None


def FunEliminateDefer(fun: cwast.DefFun):
    _EliminateDeferRecursively(fun, [])


def FunCanonicalizeRemoveStmtCond(fun: cwast.DefFun):
    """Convert StmtCond to nested StmtIf"""
    def replacer(node, _parent) -> Optional[list[Any]]:
        if not isinstance(node, cwast.StmtCond):
            return None
        out = []
        for case in reversed(node.cases):
            assert isinstance(case, cwast.Case)
            if CondIsAlwaysTrue(case.cond):
                assert not out, f"'case true' shold be the last case in a cond statement"
                out = case.body
            else:
                out = [cwast.StmtIf(case.cond, case.body,
                                    out, x_srcloc=case.x_srcloc)]
        return out

    cwast.MaybeReplaceAstRecursivelyWithParentPost(fun, replacer)


def FunAddMissingReturnStmts(fun: cwast.DefFun):

    result_ct = fun.x_type.result_type()
    if not result_ct.get_unwrapped().is_void():
        return

    sl = fun.x_srcloc
    if fun.body:
        last = fun.body[-1]
        if isinstance(last, cwast.StmtReturn):
            return
        sl = last.x_srcloc
    void_expr = cwast.ValVoid(x_srcloc=sl, x_type=result_ct)

    fun.body.append(cwast.StmtReturn(void_expr, x_srcloc=sl, x_target=fun))


def _IdNodeFromDefFun(fun: cwast.DefFun) -> cwast.Id:
    return cwast.Id(fun.name, x_srcloc=fun.x_srcloc, x_symbol=fun, x_type=fun.x_type, x_eval=eval.EvalFunAddr(fun))


def _MakeInitFiniFun(name: str, callees: list[cwast.DefFun], fun_ct, void_ct) -> cwast.DefFun:
    sl = cwast.SRCLOC_GENERATED
    at = cwast.TypeAuto(x_srcloc=sl, x_type=void_ct)
    out = cwast.DefFun(cwast.NAME(name), [], at, [],
                       cdecl=True, x_type=fun_ct, x_srcloc=sl)
    void_expr = cwast.ValVoid(x_srcloc=sl, x_type=void_ct)
    for c in callees:
        if c.x_type != fun_ct:
            cwast.CompilerError(
                c.x_srcloc, f"init function {c.name} must have type 'void ()'")
        call = cwast.ExprCall(_IdNodeFromDefFun(
            c), [], x_srcloc=cwast.SRCLOC_GENERATED, x_type=void_ct)
        out.body.append(cwast.StmtExpr(call, x_srcloc=cwast.SRCLOC_GENERATED))
    out.body.append(cwast.StmtReturn(void_expr, x_srcloc=sl, x_target=out))
    return out


def MakeInitFun(name: str, mod_topo_order: list[cwast.DefMod], tc: type_corpus.TypeCorpus) -> cwast.DefFun:
    fun_ct = tc.InsertFunType([], tc.get_void_canon_type())
    void_ct = tc.get_void_canon_type()
    lst: list[cwast.DefFun] = []
    for mod in mod_topo_order:
        seen_one = False
        for fun in mod.body_mod:
            if isinstance(fun, cwast.DefFun) and fun.init:
                lst.append(fun)
                if seen_one:
                    cwast.CompilerError(
                        fun.x_srcloc, f"multiple init functions in module {mod.name}")
                seen_one = True
    return _MakeInitFiniFun(name, lst, fun_ct, void_ct)


def MakeFiniFun(name: str, mod_topo_order: list[cwast.DefMod], tc: type_corpus.TypeCorpus) -> cwast.DefFun:
    fun_ct = tc.InsertFunType([], tc.get_void_canon_type())
    void_ct = tc.get_void_canon_type()
    lst: list[cwast.DefFun] = []
    for mod in mod_topo_order:
        seen_one = False
        for fun in reversed(mod.body_mod):
            if isinstance(fun, cwast.DefFun) and fun.fini:
                lst.append(fun)
                if seen_one:
                    cwast.CompilerError(
                        fun.x_srcloc, f"multiple fini functions in module {mod.name}")
                seen_one = True
    return _MakeInitFiniFun(name, lst, fun_ct, void_ct)
